### DISCUSSION TITLE
Defining two phase material for crystal plasticity

### DISCUSSION THREAD

**Discussion by Aayush-T** [idaholab/moose#17014] [Q&A Modules: Solid mechanics] [answered]
Hi everyone,
I wanted to understand how to define a dual-phase material for crystal plasticity that can be used to find stresses in the A and B phases as well as combined stress and strain across the whole domain. I can see there are files such as TwoPhaseStressMaterial.C and MultiPhaseStressMaterial.C, but unable to apply them for a randomly oriented Voronoi tesselation mesh. I want to do it for Ti64.

**Comment by aeslaughter** [comment]
@sapitts Can you help with this?

**Reply by sapitts** [reply]
Hi @Aayush-T, I'm always glad to hear of more folks doing crystal plasticity work! Could you give me a few more details about your plans please? E.g. will you explicitly mesh the two different phases as separate material blocks? Are you planning to use a particular crystal plasticity material model or a more general homogeneous plasticity model? These answers will help me better know what capabilities in MOOSE to suggest for your simulation

**Reply by Aayush-T** [reply]
Hi, thanks for the reply.
I would like to explicitly mesh the two different phases as separate material blocks. I am planning to use the available Kalidindi model in MOOSE without homogenization. I want to use the full field.

**Reply by sapitts** [reply]
In that case I'd recommend using the crystal plasticity model on each of the blocks, with the appropriate different properties for the different phases, something along the lines of the following. Key to your case will be the use of the elasticity tensor: here I've used the example of `ComputeElasticityTensorConstantRotationCP`, and the `euler_angle` or 'rotation_matrix' parameters can be used to set the rotation for the different phases in the Voronoi tessellation. As an example for a simple simulation with only two phases:
```
[./elasticity_tensor_phaseA]
  type = ComputeElasticityTensorConstantRotationCP
  C_ijkl = '1.684e5 1.214e5 1.214e5 1.684e5 1.214e5 1.684e5 0.754e5 0.754e5 0.754e5'
  fill_method = symmetric9
  euler_angle_1 = {phaseA}
  euler_angle_2 = {phaseA}
  euler_angle_3 = {phaseA}
[../]
[./stress_phaseA]
  type = ComputeCrystalPlasticityStress
  crystal_plasticity_update_model = 'trial_xtalpl_phaseA'
  block = phaseA
[../]
[./trial_xtalpl_phaseA]
  type = CrystalPlasticityKalidindiUpdate
  number_slip_systems = 12
  slip_sys_file_name = input_slip_sys.txt
  ao = {phaseA}
  gss_a = {phaseA}
  gss_initial = {phaseA}
  h = {phaseA}
  r = {phaseA}
  t_sat = {phaseA}
  xm= {phaseA}
  tan_mod_type = exact
[../]

[./elasticity_tensor_phaseB]
  type = ComputeElasticityTensorConstantRotationCP
  C_ijkl = '1.684e5 1.214e5 1.214e5 1.684e5 1.214e5 1.684e5 0.754e5 0.754e5 0.754e5'
  fill_method = symmetric9
  euler_angle_1 = {phaseB}
  euler_angle_2 = {phaseB}
  euler_angle_3 = {phaseB}
[../]
[./stressB]
  type = ComputeCrystalPlasticityStress
  crystal_plasticity_update_model = 'trial_xtalpl_phaseB'
  block = phaseB
[../]
[./trial_xtalpl_phaseB]
  type = CrystalPlasticityKalidindiUpdate
  number_slip_systems = 12
  slip_sys_file_name = input_slip_sys.txt
  ao = {phaseB}
  gss_a = {phaseB}
  gss_initial = {phaseB}
  h = {phaseB}
  r = {phaseB}
  t_sat = {phaseB}
  xm= {phaseB}
  tan_mod_type = exact
[../]
```
where the values in the curly braces should be set to the proper values for each phase. Then with a proper set of postprocessors you can extract the stress and strain measures across the domain. With these crystal plasticity models, I'd recommend using the Lagrangian strain and second Piola-Kirchoff stress measures; some of the regression tests use postprocessors to output these quantities as a reference example for you.

 As you've probably seen, the documentation for these classes are here:
  - [ComputeElasticityTensorConstantRotationCP](https://mooseframework.inl.gov/source/materials/crystal_plasticity/ComputeElasticityTensorConstantRotationCP.html)
  - [ComputeCrystalPlasticityStress](https://mooseframework.inl.gov/source/materials/crystal_plasticity/ComputeCrystalPlasticityStress.html)
  - [CrystalPlasticityKalidindiUpdate](https://mooseframework.inl.gov/source/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.html)

Do you have an example input file snippet you'd be willing to share, or a schematic?

**Reply by Aayush-T** [reply]
Hi, since they don't allow *.i attachments, I am changing it to *.txt to attach.
The modifications in Euler angles that you see are due to the reason that I used 2-grains in 2 phases for simplicity of mesh. I also tried this input file with 8 grains, 2 phase Voronoi tessellation mesh, but the same errors were observed. (Euler angles were defined with the help of input txt file here). 
[made_file_updated.txt](https://github.com/idaholab/moose/files/5995826/made_file_updated.txt)
The error message is also attached below.
*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Material property 'stress', requested by 'TM_TensorMechanics0' is not defined on block face1_QUAD4

**Reply by sapitts** [reply]
Remove the instances of `FiniteStrainCrystalPlasticity` from your input file: that's an older version of the crystal plasticity code that is no recommend for use. I also highly recommend the use of the [TensorMechanics MasterAction](https://mooseframework.inl.gov/syntax/Modules/TensorMechanics/Master/index.html) which will set up your displacement variables and strain calculator in a sane manner.

Take a look at some of the regression test input files provided in the repository, such as [update_method_011orientation.i](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/test/tests/crystal_plasticity/stress_update_material_based/update_method_011orientation.i) and others like it in the `tensor_mechanics/test/tests/crystal_plasticity/stress_update_material_based` directory

**Reply by Aayush-T** [reply]
Very thanks, will do the necessary changes and update to you.

**Reply by Aayush-T** [reply]
Hi, so, I tried all these changes, but again the problem persists. It asks for extra definitions. I am attaching the error snippet here too. Also, I have been recommended UO based model as the latest. So, could you advise me on using continue using KalidindiUpdate, or can use UO based crysp model. 
![image](https://user-images.githubusercontent.com/78248963/108763188-17747c00-7577-11eb-94c8-732410ebdb67.png) Also, attaching all the files related to my model as maybe I am missing something.
[crysp_multiphase.zip](https://github.com/idaholab/moose/files/6024913/crysp_multiphase.zip)

Thanks and Regards,
Aayush Trivedi

**Comment by sapitts** [comment, marked as answer]
Here's a minimum working example input file (that uses the same text files as the regression tests in this [folder](https://github.com/idaholab/moose/tree/next/modules/tensor_mechanics/test/tests/crystal_plasticity) ) for a two grain crystal plasticity problem using square shaped grains. The key correction made from the last attached input file is the removal of the `ComputeDeformGradBasedStress` input file section: the `ComputeCrystalPlasticityStress` already computes the stress. Also make sure that the different elasticity tensor and `ComputeCrystalPlasticityStress` input file sections for the different phases are restricted to different mesh blocks (e.g. `block = 1` and `block = 2` in the example below)

```
[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Mesh]
  [two_block]
    type = CartesianMeshGenerator
    dim = 3
    dx = '1.0 1.0'
    dy = '1.0'
    dz = '1.0'
    ix = '2 2'
    iy = '2'
    iz = '2'
    subdomain_id = '1 2'
  []
[]

[AuxVariables]
  [pk2]
    order = CONSTANT
    family = MONOMIAL
  []
  [fp_zz]
    order = CONSTANT
    family = MONOMIAL
  []
  [lagrangian_strain_zz]
    order = CONSTANT
    family = MONOMIAL
  []
  [gss]
    order = CONSTANT
    family = MONOMIAL
  []
  [slip_increment]
   order = CONSTANT
   family = MONOMIAL
  []
[]

[Modules/TensorMechanics/Master/all]
  strain = FINITE
  add_variables = true
  generate_output = stress_zz
[]

[AuxKernels]
  [pk2]
    type = RankTwoAux
    variable = pk2
    rank_two_tensor = pk2
    index_j = 2
    index_i = 2
    execute_on = timestep_end
  []
  [fp_zz]
    type = RankTwoAux
    variable = fp_zz
    rank_two_tensor = fp
    index_j = 2
    index_i = 2
    execute_on = timestep_end
  []
  [lagrangian_strain_zz]
    type = RankTwoAux
    variable = lagrangian_strain_zz
    rank_two_tensor = total_lagrangian_strain
    index_j = 2
    index_i = 2
    execute_on = timestep_end
  []
  [gss]
    type = MaterialStdVectorAux
    variable = gss
    property = slip_system_resistance
    index = 0
    execute_on = timestep_end
  []
  [slip_inc]
   type = MaterialStdVectorAux
   variable = slip_increment
   property = plastic_slip_increment
   index = 0
   execute_on = timestep_end
  []
[]

[BCs]
  [symmy]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0
  []
  [symmx]
    type = DirichletBC
    variable = disp_x
    boundary = left
    value = 0
  []
  [symmz]
    type = DirichletBC
    variable = disp_z
    boundary = back
    value = 0
  []
  [tdisp]
    type = FunctionDirichletBC
    variable = disp_z
    boundary = front
    function = '0.01*t'
  []
[]

[Materials]
  [elasticity_tensor_phaseA]
    type = ComputeElasticityTensorConstantRotationCP
    C_ijkl = '1.684e5 1.214e5 1.214e5 1.684e5 1.214e5 1.684e5 0.754e5 0.754e5 0.754e5'
    fill_method = symmetric9
    # base_name = A
    block = 1
  []
  [stress_phaseA]
    type = ComputeCrystalPlasticityStress
    crystal_plasticity_update_model = 'trial_xtalpl_phaseA'
    # base_name = A
    block = 1
  []
  [trial_xtalpl_phaseA]
    type = CrystalPlasticityKalidindiUpdate
    number_slip_systems = 12
    slip_sys_file_name = input_slip_sys.txt
    tan_mod_type = exact
    # base_name = A
    block = 1
  []
  [elasticity_tensor_phaseB]
    type = ComputeElasticityTensorConstantRotationCP
    C_ijkl = '1.684e5 1.214e5 1.214e5 1.684e5 1.214e5 1.684e5 0.754e5 0.754e5 0.754e5'
    fill_method = symmetric9
    euler_angle_1 = 120.0
    euler_angle_2 = 125.264
    euler_angle_3 =  45.0
    # base_name = B
    block = 2
  []
  [stress_phaseB]
    type = ComputeCrystalPlasticityStress
    crystal_plasticity_update_model = 'trial_xtalpl_phaseB'
    # base_name = B
    block = 2
  []
  [trial_xtalpl_phaseB]
    type = CrystalPlasticityKalidindiUpdate
    number_slip_systems = 12
    slip_sys_file_name = input_slip_sys.txt
    tan_mod_type = exact
    # base_name = B
    block = 2
  []
[]

[Postprocessors]
  [stress_zz]
    type = ElementAverageValue
    variable = stress_zz
  []
  [pk2]
   type = ElementAverageValue
   variable = pk2
  []
  [fp_zz]
    type = ElementAverageValue
    variable = fp_zz
  []
  [lagrangian_strain_zz]
    type = ElementAverageValue
    variable = lagrangian_strain_zz
  []
  [gss]
    type = ElementAverageValue
    variable = gss
  []
  [slip_increment]
   type = ElementAverageValue
   variable = slip_increment
  []
[]

[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient
  solve_type = 'PJFNK'

  petsc_options_iname = '-pc_type -pc_asm_overlap -sub_pc_type -ksp_type -ksp_gmres_restart'
  petsc_options_value = ' asm      2              lu            gmres     200'
  nl_abs_tol = 1e-10
  nl_rel_tol = 1e-10
  nl_abs_step_tol = 1e-10

  dt = 0.05
  dtmin = 0.01
  dtmax = 10.0
  num_steps = 100
[]

[Outputs]
  exodus = true
[]
```

Note that the Kalidindi model (in both the material-based stress update and in the user-object based model) is formulated for a 12 slip system FCC crystal. If you want to use 24 slip systems in a BCC lattice you may want to consider changing to a different crystal plasticity model once you have the proof-of-concept input file up and running.

**Reply by sapitts** [reply]
If you are willing to accommodate a few input file and base class changes in the next six months or so, I'd recommend using the material based `Update` version over the user-object based version. @dewenyushu and I are actively developing the material-based `Update` version, and the INL team is transistioning away from the user-object based version. Both versions will work well, however, and at present more constitutive models are available in the user-object version (e.g. Voce model). We are working on transitioning these models to the material-based version, and any new constitutive models we add will be only in the material-based `Update` version.

The material-based `Update` version has been developed to have a complete constitutive model within a material class. This implementation more closely aligns with the descriptions of a constitutive model that one might read in a journal article. This collection of all the constitutive model details within a material class has a couple of main benefits:
  - More direct storage of state variables (e.g. slip system resistance), and our initial results show some speed up with the material-based version
  - Easier to modify and implement new constitutive models

Hope this helps!

**Reply by Aayush-T** [reply]
Hi, by material-based Update version, you mean CrystalPlasticityKalidindiUpdate, right? Or is there any other material-based new version? Yes, I am willing to accommodate that. Also, for the model, I wanted to use the CrystalPlasticityUObased for plasticity in dual-phase to check its stability, but it always gives similar errors as before.
[Error_file.zip](https://github.com/idaholab/moose/files/6045416/Error_file.zip)
Moreover, I was wondering if I could get the values of stresses in A and B section in KalidindiUpdateModel for two-phase that was being developed. I have found a way to get it through exodus file output but is it also possible to get it with moose postprocessors. I did it previously in my models, but there seems to be some error this time. Also, it may look incomplete in the postprocessor as I was still trying to do it.
[error_stressA.zip](https://github.com/idaholab/moose/files/6045561/error_stressA.zip)

**Reply by ngrilli** [reply]
Dear @Aayush-T and @sapitts 

My understanding is that you cannot use together 
ComputeCrystalPlasticityStress and FiniteStrainUObasedCP on the same block
because they both declare stress as material property.

I think FiniteStrainUObasedCP is not needed anymore. Isn't it? 
That should be the reason you get that error.

I am really interested to see the new implementation 
of the crystal plasticity model in MOOSE.

Best Regards,
Nicolò Grilli
National University of Singapore

**Reply by Aayush-T** [reply]
Yes, that is right. I was initiating it twice. I removed the ComputeCrystalPlasticityStress to utilize FiniteStrainUObasedCP. It solves the issue, thank you. Also, the issue for stress values can be resolved by using exodus file as I was getting the average value for overall elements in the MOOSE output, I did try to get the A_stress values and I failed the model there. It won't affect anything else though.
[./A_stress_yy]
     type = RankTwoAux
     variable =A_stress_yy
     rank_two_tensor = A_stress
     index_j = 1
     index_i = 1
     execute_on = timestep_end
     block = 'face1_QUAD4'
  [../]
I used this as the Kernal. 
Moreover, I am also interested in the new implementation of the crystal plasticity model in MOOSE.

**Reply by sapitts** [reply]
Hi @Aayush-T  you can also get the values of the stresses in the different blocks by "block restricting" postprocessors. If you wanted the average stress value in the entire domain and two individual blocks, something like the following ought to create those three values:

```
[Postprocessors]
  [pk2_entire_domain]
    type = ElementAverageValue
    variable = pk2
    block = ' 1 2' ## the default is all blocks in a mesh, so this line isn't strictly necessary here
  []
  [./pk2_block1]
   type = ElementAverageValue
   variable = pk2
   block = 1
  []
  [pk2_block2]
   type = ElementAverageValue
   variable = pk2
   block = 2
  []
```

From the input file you attached in a previous post it looks like you were quite close to this format with the `stress_B` postprocessor.

Hope this helps!

**Comment by sapitts** [comment]
@ngrilli and @Aayush-T I'm glad to hear that you are both interested in the material based CrystalPlasticity{...}Update classes. We're working to retain the flexibility provided by the user-object (the `FiniteStrainUObasedCP` version) while giving the code a more "material-scientist friendly" structure and format.  As @dewenyushu and I get working examples added to the MOOSE repository we will add that information to the MOOSE newsletter
